File: python/neon/decoder.py
"""
NEON Decoder Implementation
"""

import re
import time
from typing import Any, Callable, Dict, List, Optional, Tuple

from .types import (
    NEON_EXPANSIONS,
    NEON_SYMBOLS,
    NeonArray,
    NeonDecodeOptions,
    NeonObject,
    NeonStats,
    NeonSyntaxError,
    NeonValue,
)


def expand_number(s: str) -> float:
    """Expand abbreviated numbers."""
    if not s:
        return 0.0

    s = s.strip()

    if s.endswith("T"):
        return float(s[:-1]) * 1_000_000_000_000
    if s.endswith("B"):
        return float(s[:-1]) * 1_000_000_000
    if s.endswith("M"):
        return float(s[:-1]) * 1_000_000
    if s.endswith("K"):
        return float(s[:-1]) * 1_000

    # Leading dot decimal
    if s.startswith("."):
        return float("0" + s)
    if s.startswith("-."):
        return float("-0" + s[1:])

    return float(s)


class NeonLexer:
    """Tokenizer for NEON format."""

    def __init__(self, input_str: str):
        self.input = input_str
        self.pos = 0
        self.line = 1
        self.column = 1
        self.tokens: List[Dict[str, Any]] = []

    def tokenize(self) -> List[Dict[str, Any]]:
        """Tokenize the input string."""
        while self.pos < len(self.input):
            self._scan_token()

        self._add_token("EOF", "", "")
        return self.tokens

    def _scan_token(self) -> None:
        char = self.input[self.pos]

        # Skip whitespace (except newlines)
        if char in " \t\r":
            self._advance()
            return

        # Newline
        if char == "\n":
            self._add_token("NEWLINE", "\n", "\n")
            self._advance()
            self.line += 1
            self.column = 1
            return

        # Check indentation at start of line
        if self.column == 1 and char == " ":
            indent = ""
            while self.pos < len(self.input) and self.input[self.pos] == " ":
                indent += " "
                self._advance()
            if indent:
                self._add_token("INDENT", indent, indent)
            return

        # Symbols
        symbol_map = {
            NEON_SYMBOLS["ARRAY"]: "ARRAY_START",
            NEON_SYMBOLS["SCHEMA"]: "SCHEMA_START",
            NEON_SYMBOLS["OBJECT"]: "OBJECT_START",
            ":": "COLON",
            ",": "COMMA",
            NEON_SYMBOLS["LIST_ITEM"]: "LIST_ITEM",
        }

        if char in symbol_map:
            self._add_token(symbol_map[char], char, char)
            self._advance()
            return

        # Boolean shortcuts
        if char == NEON_SYMBOLS["TRUE"] and self._is_word_boundary(self.pos + 1):
            self._add_token("BOOLEAN", "true", char)
            self._advance()
            return

        if char == NEON_SYMBOLS["FALSE"] and self._is_word_boundary(self.pos + 1):
            self._add_token("BOOLEAN", "false", char)
            self._advance()
            return

        # Null shortcut
        if char == NEON_SYMBOLS["NULL"] and self._is_word_boundary(self.pos + 1):
            self._add_token("NULL", "null", char)
            self._advance()
            return

        # Quoted string
        if char == '"':
            self._scan_quoted_string()
            return

        # Number
        if self._is_number_start(char):
            self._scan_number()
            return

        # Unquoted string
        self._scan_unquoted_string()

    def _scan_quoted_string(self) -> None:
        self._advance()  # Skip opening quote
        start = self.pos
        value = ""

        while self.pos < len(self.input) and self.input[self.pos] != '"':
            if self.input[self.pos] == "\\":
                self._advance()
                if self.pos < len(self.input):
                    escape_char = self.input[self.pos]
                    escape_map = {"n": "\n", "r": "\r", "t": "\t", '"': '"', "\\": "\\"}
                    value += escape_map.get(escape_char, escape_char)
            else:
                value += self.input[self.pos]
            self._advance()

        raw = self.input[start : self.pos]
        self._advance()  # Skip closing quote
        self._add_token("STRING", value, f'"{raw}"')

    def _scan_number(self) -> None:
        start = self.pos

        # Optional negative sign
        if self.input[self.pos] == "-":
            self._advance()

        # Leading dot
        if self.pos < len(self.input) and self.input[self.pos] == ".":
            self._advance()

        # Digits
        while self.pos < len(self.input) and self.input[self.pos].isdigit():
            self._advance()

        # Decimal part
        if self.pos < len(self.input) and self.input[self.pos] == ".":
            self._advance()
            while self.pos < len(self.input) and self.input[self.pos].isdigit():
                self._advance()

        # Suffix (K, M, B, T)
        if self.pos < len(self.input) and self.input[self.pos] in "KMBT":
            self._advance()

        raw = self.input[start : self.pos]
        value = str(expand_number(raw))
        self._add_token("NUMBER", value, raw)

    def _scan_unquoted_string(self) -> None:
        start = self.pos
        delimiters = ' \t\n\r:,#@$~^>"'

        while self.pos < len(self.input) and self.input[self.pos] not in delimiters:
            self._advance()

        raw = self.input[start : self.pos]

        # Check for keywords
        if raw == "null":
            self._add_token("NULL", "null", raw)
        elif raw == "true":
            self._add_token("BOOLEAN", "true", raw)
        elif raw == "false":
            self._add_token("BOOLEAN", "false", raw)
        else:
            # Convert underscores to spaces
            value = raw.replace("_", " ")
            self._add_token("STRING", value, raw)

    def _is_number_start(self, char: str) -> bool:
        return char.isdigit() or char == "-" or char == "."

    def _is_word_boundary(self, pos: int) -> bool:
        if pos >= len(self.input):
            return True
        return self.input[pos] in ' \t\n\r:,#@$~^>"'

    def _advance(self) -> None:
        self.pos += 1
        self.column += 1

    def _add_token(self, token_type: str, value: str, raw: str) -> None:
        self.tokens.append({
            "type": token_type,
            "value": value,
            "raw": raw,
            "line": self.line,
            "column": self.column,
        })


class NeonParser:
    """Parser for NEON format."""

    def __init__(self, options: Optional[NeonDecodeOptions] = None):
        self.options = options or NeonDecodeOptions()
        self.tokens: List[Dict[str, Any]] = []
        self.current = 0
        self.depth = 0

    def parse(self, tokens: List[Dict[str, Any]]) -> NeonValue:
        """Parse tokens into a Python value."""
        self.tokens = tokens
        self.current = 0
        self.depth = 0

        self._skip_newlines()

        if self._is_at_end():
            return None

        return self._parse_value()

    def _parse_value(self) -> NeonValue:
        self.depth += 1

        if self.depth > self.options.max_depth:
            raise NeonSyntaxError(
                "MAX_DEPTH",
                f"Maximum depth of {self.options.max_depth} exceeded",
                self._peek().get("line", 0),
                self._peek().get("column", 0),
            )

        token = self._peek()
        if not token or token["type"] == "EOF":
            self.depth -= 1
            return None

        result: NeonValue = None

        # Check for named array first (e.g., "users#5^...")
        if token["type"] == "STRING":
            next_token = self._peek_next()
            if next_token and next_token["type"] == "ARRAY_START":
                result = self._parse_named_array()
                self.depth -= 1
                return result

        token_type = token["type"]

        if token_type == "NULL":
            self._advance()
            result = None
        elif token_type == "BOOLEAN":
            self._advance()
            result = token["value"] == "true"
        elif token_type == "NUMBER":
            self._advance()
            num = float(token["value"])
            result = int(num) if num.is_integer() else num
        elif token_type == "STRING":
            self._advance()
            result = self._expand_abbreviation(token["value"])
        elif token_type == "OBJECT_START":
            result = self._parse_object()
        elif token_type == "ARRAY_START":
            result = self._parse_array()
        else:
            self._advance()
            result = token["value"]

        self.depth -= 1
        return result

    def _parse_object(self) -> NeonObject:
        self._expect("OBJECT_START")
        obj: NeonObject = {}

        while not self._is_at_end() and not self._check("NEWLINE") and not self._check("EOF"):
            key_token = self._peek()
            if not key_token or key_token["type"] != "STRING":
                break

            self._advance()
            key = self._expand_abbreviation(key_token["value"])

            if self._check("COLON"):
                self._advance()

                if self._check("ARRAY_START"):
                    obj[key] = self._parse_array()
                elif self._check("NEWLINE"):
                    self._skip_newlines()
                    if self._check("INDENT"):
                        self._advance()
                        obj[key] = self._parse_value()
                    else:
                        obj[key] = None
                else:
                    obj[key] = self._parse_value()
            elif self._check("ARRAY_START"):
                obj[key] = self._parse_array()
            else:
                break

        return obj

    def _parse_array(self) -> NeonArray:
        self._expect("ARRAY_START")
        arr: NeonArray = []

        # Get length
        length_token = self._peek()
        if not length_token or length_token["type"] != "NUMBER":
            raise NeonSyntaxError(
                "SYNTAX",
                "Expected array length",
                length_token.get("line", 0) if length_token else 0,
                length_token.get("column", 0) if length_token else 0,
            )

        self._advance()
        length = int(float(length_token["value"]))

        if length == 0:
            return []

        # Check for schema
        schema: Optional[List[str]] = None
        if self._check("SCHEMA_START"):
            self._advance()
            schema = self._parse_schema()

        # Check for inline values
        if not self._check("NEWLINE") and not self._check("EOF"):
            # Inline primitive array
            while len(arr) < length and not self._is_at_end():
                if self._check("NEWLINE") or self._check("EOF"):
                    break
                arr.append(self._parse_value())

            return arr

        # Multiline array
        if schema:
            return self._parse_tabular_rows(length, schema)
        else:
            return self._parse_list_rows(length)

    def _parse_named_array(self) -> NeonObject:
        """Parse a named array like 'users#3^id,name'."""
        name_token = self._peek()
        self._advance()  # Skip name
        name = name_token["value"]

        arr = self._parse_array()
        return {name: arr}

    def _parse_schema(self) -> List[str]:
        """Parse schema fields."""
        fields: List[str] = []

        while not self._is_at_end():
            token = self._peek()
            if token["type"] != "STRING":
                break

            self._advance()
            fields.append(self._expand_abbreviation(token["value"]))

            if self._check("COMMA"):
                self._advance()
            else:
                break

        return fields

    def _parse_tabular_rows(self, length: int, fields: List[str]) -> NeonArray:
        """Parse tabular array rows."""
        result: NeonArray = []

        for _ in range(length):
            self._skip_newlines()

            if self._check("INDENT"):
                self._advance()

            obj: NeonObject = {}
            for i, field in enumerate(fields):
                if self._is_at_end() or self._check("NEWLINE"):
                    break
                obj[field] = self._parse_value()

            result.append(obj)

        return result

    def _parse_list_rows(self, length: int) -> NeonArray:
        """Parse list-style array rows."""
        result: NeonArray = []

        for _ in range(length):
            self._skip_newlines()

            if self._check("INDENT"):
                self._advance()

            if self._check("LIST_ITEM"):
                self._advance()

            result.append(self._parse_value())

        return result

    def _expand_abbreviation(self, value: str) -> str:
        """Expand field abbreviation if enabled."""
        if self.options.expand_abbreviations:
            return NEON_EXPANSIONS.get(value, value)
        return value

    def _peek(self) -> Optional[Dict[str, Any]]:
        if self.current >= len(self.tokens):
            return None
        return self.tokens[self.current]

    def _peek_next(self) -> Optional[Dict[str, Any]]:
        if self.current + 1 >= len(self.tokens):
            return None
        return self.tokens[self.current + 1]

    def _advance(self) -> Optional[Dict[str, Any]]:
        if not self._is_at_end():
            self.current += 1
        return self.tokens[self.current - 1] if self.current > 0 else None

    def _check(self, token_type: str) -> bool:
        if self._is_at_end():
            return False
        token = self._peek()
        return token is not None and token["type"] == token_type

    def _expect(self, token_type: str) -> Dict[str, Any]:
        if self._check(token_type):
            return self._advance()
        token = self._peek()
        raise NeonSyntaxError(
            "SYNTAX",
            f"Expected {token_type}",
            token.get("line", 0) if token else 0,
            token.get("column", 0) if token else 0,
        )

    def _is_at_end(self) -> bool:
        token = self._peek()
        return token is None or token["type"] == "EOF"

    def _skip_newlines(self) -> None:
        while self._check("NEWLINE"):
            self._advance()


class NeonDecoder:
    """NEON format decoder."""

    def __init__(self, options: Optional[NeonDecodeOptions] = None):
        self.options = options or NeonDecodeOptions()
        self.stats = NeonStats()

    def decode(self, input_str: str) -> NeonValue:
        """Decode a NEON string to Python value."""
        start_time = time.perf_counter()

        # Handle empty input
        if not input_str or not input_str.strip():
            return None

        # Tokenize
        lexer = NeonLexer(input_str)
        tokens = lexer.tokenize()

        # Parse
        parser = NeonParser(self.options)
        result = parser.parse(tokens)

        end_time = time.perf_counter()

        # Calculate stats
        import json
        output_str = json.dumps(result) if result is not None else ""
        self.stats = NeonStats(
            input_size=len(input_str),
            output_size=len(output_str),
            decode_time_ms=(end_time - start_time) * 1000,
        )

        return result

    def get_stats(self) -> NeonStats:
        """Get statistics from the last decode operation."""
        return self.stats


def decode(input_str: str, options: Optional[NeonDecodeOptions] = None) -> NeonValue:
    """Decode a NEON string to Python value."""
    decoder = NeonDecoder(options)
    return decoder.decode(input_str)


def decode_stream(
    input_str: str,
    callback: Callable[[NeonValue, int], None],
    options: Optional[NeonDecodeOptions] = None,
) -> None:
    """Decode NEON and process records via callback."""
    result = decode(input_str, options)

    if isinstance(result, dict):
        for key, value in result.items():
            if isinstance(value, list):
                for i, item in enumerate(value):
                    callback(item, i)
                return

    if isinstance(result, list):
        for i, item in enumerate(result):
            callback(item, i)

File: rust/src/decoder.rs
//! NEON Decoder Implementation

use crate::error::{NeonError, Result};
use crate::types::{symbols, get_expansions, NeonDecodeOptions, NeonStats, Token, TokenType};
use serde_json::{Map, Number, Value};
use std::time::Instant;

/// Expand abbreviated numbers
fn expand_number(s: &str) -> f64 {
    let s = s.trim();
    if s.is_empty() {
        return 0.0;
    }

    if s.ends_with('T') {
        return s[..s.len() - 1].parse::<f64>().unwrap_or(0.0) * 1_000_000_000_000.0;
    }
    if s.ends_with('B') {
        return s[..s.len() - 1].parse::<f64>().unwrap_or(0.0) * 1_000_000_000.0;
    }
    if s.ends_with('M') {
        return s[..s.len() - 1].parse::<f64>().unwrap_or(0.0) * 1_000_000.0;
    }
    if s.ends_with('K') {
        return s[..s.len() - 1].parse::<f64>().unwrap_or(0.0) * 1_000.0;
    }

    // Leading dot
    if s.starts_with('.') {
        return format!("0{}", s).parse().unwrap_or(0.0);
    }
    if s.starts_with("-.") {
        return format!("-0{}", &s[1..]).parse().unwrap_or(0.0);
    }

    s.parse().unwrap_or(0.0)
}

/// Lexer for NEON format
struct Lexer {
    input: Vec<char>,
    pos: usize,
    line: usize,
    column: usize,
    tokens: Vec<Token>,
}

impl Lexer {
    fn new(input: &str) -> Self {
        Self {
            input: input.chars().collect(),
            pos: 0,
            line: 1,
            column: 1,
            tokens: Vec::new(),
        }
    }

    fn tokenize(&mut self) -> Vec<Token> {
        while self.pos < self.input.len() {
            self.scan_token();
        }

        self.add_token(TokenType::Eof, String::new(), String::new());
        std::mem::take(&mut self.tokens)
    }

    fn scan_token(&mut self) {
        let c = self.input[self.pos];

        // Skip whitespace (except newlines)
        if c == ' ' || c == '\t' || c == '\r' {
            self.advance();
            return;
        }

        // Newline
        if c == '\n' {
            self.add_token(TokenType::Newline, "\n".to_string(), "\n".to_string());
            self.advance();
            self.line += 1;
            self.column = 1;
            return;
        }

        // Symbols
        match c {
            c if c == symbols::ARRAY => {
                self.add_token(TokenType::ArrayStart, c.to_string(), c.to_string());
                self.advance();
                return;
            }
            c if c == symbols::SCHEMA => {
                self.add_token(TokenType::SchemaStart, c.to_string(), c.to_string());
                self.advance();
                return;
            }
            c if c == symbols::OBJECT => {
                self.add_token(TokenType::ObjectStart, c.to_string(), c.to_string());
                self.advance();
                return;
            }
            c if c == symbols::COLON => {
                self.add_token(TokenType::Colon, c.to_string(), c.to_string());
                self.advance();
                return;
            }
            c if c == symbols::COMMA => {
                self.add_token(TokenType::Comma, c.to_string(), c.to_string());
                self.advance();
                return;
            }
            c if c == symbols::LIST_ITEM => {
                self.add_token(TokenType::ListItem, c.to_string(), c.to_string());
                self.advance();
                return;
            }
            _ => {}
        }

        // Boolean shortcuts
        if c == symbols::TRUE && self.is_word_boundary(self.pos + 1) {
            self.add_token(TokenType::Boolean, "true".to_string(), c.to_string());
            self.advance();
            return;
        }
        if c == symbols::FALSE && self.is_word_boundary(self.pos + 1) {
            self.add_token(TokenType::Boolean, "false".to_string(), c.to_string());
            self.advance();
            return;
        }
        if c == symbols::NULL && self.is_word_boundary(self.pos + 1) {
            self.add_token(TokenType::Null, "null".to_string(), c.to_string());
            self.advance();
            return;
        }

        // Quoted string
        if c == '"' {
            self.scan_quoted_string();
            return;
        }

        // Number
        if self.is_number_start(c) {
            self.scan_number();
            return;
        }

        // Unquoted string
        self.scan_unquoted_string();
    }

    fn scan_quoted_string(&mut self) {
        self.advance(); // Skip opening quote
        let start = self.pos;
        let mut value = String::new();

        while self.pos < self.input.len() && self.input[self.pos] != '"' {
            if self.input[self.pos] == '\\' {
                self.advance();
                if self.pos < self.input.len() {
                    let escape_char = self.input[self.pos];
                    match escape_char {
                        'n' => value.push('\n'),
                        'r' => value.push('\r'),
                        't' => value.push('\t'),
                        '"' => value.push('"'),
                        '\\' => value.push('\\'),
                        _ => value.push(escape_char),
                    }
                }
            } else {
                value.push(self.input[self.pos]);
            }
            self.advance();
        }

        let raw: String = self.input[start..self.pos].iter().collect();
        self.advance(); // Skip closing quote
        self.add_token(TokenType::String, value, format!("\"{}\"", raw));
    }

    fn scan_number(&mut self) {
        let start = self.pos;

        // Optional negative sign
        if self.input[self.pos] == '-' {
            self.advance();
        }

        // Leading dot
        if self.pos < self.input.len() && self.input[self.pos] == '.' {
            self.advance();
        }

        // Digits
        while self.pos < self.input.len() && self.input[self.pos].is_ascii_digit() {
            self.advance();
        }

        // Decimal part
        if self.pos < self.input.len() && self.input[self.pos] == '.' {
            self.advance();
            while self.pos < self.input.len() && self.input[self.pos].is_ascii_digit() {
                self.advance();
            }
        }

        // Suffix (K, M, B, T)
        if self.pos < self.input.len() && "KMBT".contains(self.input[self.pos]) {
            self.advance();
        }

        let raw: String = self.input[start..self.pos].iter().collect();
        let value = expand_number(&raw).to_string();
        self.add_token(TokenType::Number, value, raw);
    }

    fn scan_unquoted_string(&mut self) {
        let start = self.pos;
        let delimiters = " \t\n\r:,#@$~^>\"";

        while self.pos < self.input.len() && !delimiters.contains(self.input[self.pos]) {
            self.advance();
        }

        let raw: String = self.input[start..self.pos].iter().collect();

        // Check for keywords
        match raw.as_str() {
            "null" => self.add_token(TokenType::Null, "null".to_string(), raw),
            "true" => self.add_token(TokenType::Boolean, "true".to_string(), raw),
            "false" => self.add_token(TokenType::Boolean, "false".to_string(), raw),
            _ => {
                // Convert underscores to spaces
                let value = raw.replace('_', " ");
                self.add_token(TokenType::String, value, raw);
            }
        }
    }

    fn is_number_start(&self, c: char) -> bool {
        c.is_ascii_digit() || c == '-' || c == '.'
    }

    fn is_word_boundary(&self, pos: usize) -> bool {
        if pos >= self.input.len() {
            return true;
        }
        let delimiters = " \t\n\r:,#@$~^>\"";
        delimiters.contains(self.input[pos])
    }

    fn advance(&mut self) {
        self.pos += 1;
        self.column += 1;
    }

    fn add_token(&mut self, token_type: TokenType, value: String, raw: String) {
        self.tokens.push(Token {
            token_type,
            value,
            raw,
            line: self.line,
            column: self.column,
        });
    }
}

/// Parser for NEON format
struct Parser {
    options: NeonDecodeOptions,
    tokens: Vec<Token>,
    current: usize,
    depth: usize,
}

impl Parser {
    fn new(options: NeonDecodeOptions) -> Self {
        Self {
            options,
            tokens: Vec::new(),
            current: 0,
            depth: 0,
        }
    }

    fn parse(&mut self, tokens: Vec<Token>) -> Result<Value> {
        self.tokens = tokens;
        self.current = 0;
        self.depth = 0;

        self.skip_newlines();

        if self.is_at_end() {
            return Ok(Value::Null);
        }

        self.parse_value()
    }

    fn parse_value(&mut self) -> Result<Value> {
        self.depth += 1;

        if self.depth > self.options.max_depth {
            return Err(NeonError::MaxDepth {
                depth: self.options.max_depth,
            });
        }

        let token = match self.peek() {
            Some(t) => t.clone(),
            None => {
                self.depth -= 1;
                return Ok(Value::Null);
            }
        };

        if token.token_type == TokenType::Eof {
            self.depth -= 1;
            return Ok(Value::Null);
        }

        // Check for named array
        if token.token_type == TokenType::String {
            if let Some(next) = self.peek_next() {
                if next.token_type == TokenType::ArrayStart {
                    let result = self.parse_named_array()?;
                    self.depth -= 1;
                    return Ok(result);
                }
            }
        }

        let result = match token.token_type {
            TokenType::Null => {
                self.advance();
                Value::Null
            }
            TokenType::Boolean => {
                self.advance();
                Value::Bool(token.value == "true")
            }
            TokenType::Number => {
                self.advance();
                let num: f64 = token.value.parse().unwrap_or(0.0);
                if num.fract() == 0.0 {
                    Value::Number(Number::from(num as i64))
                } else {
                    Value::Number(Number::from_f64(num).unwrap_or(Number::from(0)))
                }
            }
            TokenType::String => {
                self.advance();
                Value::String(self.expand_abbreviation(&token.value))
            }
            TokenType::ObjectStart => self.parse_object()?,
            TokenType::ArrayStart => self.parse_array()?,
            _ => {
                self.advance();
                Value::String(token.value)
            }
        };

        self.depth -= 1;
        Ok(result)
    }

    fn parse_object(&mut self) -> Result<Value> {
        self.expect(TokenType::ObjectStart)?;
        let mut obj = Map::new();

        while !self.is_at_end() && !self.check(TokenType::Newline) && !self.check(TokenType::Eof) {
            let key_token = match self.peek() {
                Some(t) if t.token_type == TokenType::String => t.clone(),
                _ => break,
            };

            self.advance();
            let key = self.expand_abbreviation(&key_token.value);

            if self.check(TokenType::Colon) {
                self.advance();

                if self.check(TokenType::ArrayStart) {
                    obj.insert(key, self.parse_array()?);
                } else if self.check(TokenType::Newline) {
                    self.skip_newlines();
                    if self.check(TokenType::Indent) {
                        self.advance();
                        obj.insert(key, self.parse_value()?);
                    } else {
                        obj.insert(key, Value::Null);
                    }
                } else {
                    obj.insert(key, self.parse_value()?);
                }
            } else if self.check(TokenType::ArrayStart) {
                obj.insert(key, self.parse_array()?);
            } else {
                break;
            }
        }

        Ok(Value::Object(obj))
    }

    fn parse_array(&mut self) -> Result<Value> {
        self.expect(TokenType::ArrayStart)?;

        let length_token = self
            .peek()
            .ok_or_else(|| NeonError::syntax("Expected array length", 0, 0))?
            .clone();

        if length_token.token_type != TokenType::Number {
            return Err(NeonError::syntax("Expected array length", 0, 0));
        }

        self.advance();
        let length: usize = length_token.value.parse().unwrap_or(0) as usize;

        if length == 0 {
            return Ok(Value::Array(Vec::new()));
        }

        // Check for schema
        let schema = if self.check(TokenType::SchemaStart) {
            self.advance();
            Some(self.parse_schema()?)
        } else {
            None
        };

        // Check for inline values
        if !self.check(TokenType::Newline) && !self.check(TokenType::Eof) {
            let mut arr = Vec::new();
            while arr.len() < length && !self.is_at_end() {
                if self.check(TokenType::Newline) || self.check(TokenType::Eof) {
                    break;
                }
                arr.push(self.parse_value()?);
            }
            return Ok(Value::Array(arr));
        }

        // Multiline array
        if let Some(fields) = schema {
            self.parse_tabular_rows(length, &fields)
        } else {
            self.parse_list_rows(length)
        }
    }

    fn parse_named_array(&mut self) -> Result<Value> {
        let name_token = self.peek().unwrap().clone();
        self.advance();
        let name = name_token.value;

        let arr = self.parse_array()?;

        let mut obj = Map::new();
        obj.insert(name, arr);
        Ok(Value::Object(obj))
    }

    fn parse_schema(&mut self) -> Result<Vec<String>> {
        let mut fields = Vec::new();

        while !self.is_at_end() {
            let token = match self.peek() {
                Some(t) if t.token_type == TokenType::String => t.clone(),
                _ => break,
            };

            self.advance();
            fields.push(self.expand_abbreviation(&token.value));

            if self.check(TokenType::Comma) {
                self.advance();
            } else {
                break;
            }
        }

        Ok(fields)
    }

    fn parse_tabular_rows(&mut self, length: usize, fields: &[String]) -> Result<Value> {
        let mut result = Vec::new();

        for _ in 0..length {
            self.skip_newlines();

            if self.check(TokenType::Indent) {
                self.advance();
            }

            let mut obj = Map::new();
            for field in fields {
                if self.is_at_end() || self.check(TokenType::Newline) {
                    break;
                }
                obj.insert(field.clone(), self.parse_value()?);
            }

            result.push(Value::Object(obj));
        }

        Ok(Value::Array(result))
    }

    fn parse_list_rows(&mut self, length: usize) -> Result<Value> {
        let mut result = Vec::new();

        for _ in 0..length {
            self.skip_newlines();

            if self.check(TokenType::Indent) {
                self.advance();
            }

            if self.check(TokenType::ListItem) {
                self.advance();
            }

            result.push(self.parse_value()?);
        }

        Ok(Value::Array(result))
    }

    fn expand_abbreviation(&self, value: &str) -> String {
        if self.options.expand_abbreviations {
            get_expansions()
                .get(value)
                .map(|s| s.to_string())
                .unwrap_or_else(|| value.to_string())
        } else {
            value.to_string()
        }
    }

    fn peek(&self) -> Option<&Token> {
        self.tokens.get(self.current)
    }

    fn peek_next(&self) -> Option<&Token> {
        self.tokens.get(self.current + 1)
    }

    fn advance(&mut self) -> Option<&Token> {
        if !self.is_at_end() {
            self.current += 1;
        }
        self.tokens.get(self.current - 1)
    }

    fn check(&self, token_type: TokenType) -> bool {
        self.peek()
            .map(|t| t.token_type == token_type)
            .unwrap_or(false)
    }

    fn expect(&mut self, token_type: TokenType) -> Result<&Token> {
        if self.check(token_type.clone()) {
            Ok(self.advance().unwrap())
        } else {
            Err(NeonError::syntax(
                format!("Expected {:?}", token_type),
                self.peek().map(|t| t.line).unwrap_or(0),
                self.peek().map(|t| t.column).unwrap_or(0),
            ))
        }
    }

    fn is_at_end(&self) -> bool {
        self.peek()
            .map(|t| t.token_type == TokenType::Eof)
            .unwrap_or(true)
    }

    fn skip_newlines(&mut self) {
        while self.check(TokenType::Newline) {
            self.advance();
        }
    }
}

/// NEON Decoder
pub struct NeonDecoder {
    options: NeonDecodeOptions,
    stats: NeonStats,
}

impl NeonDecoder {
    pub fn new(options: Option<NeonDecodeOptions>) -> Self {
        Self {
            options: options.unwrap_or_default(),
            stats: NeonStats::default(),
        }
    }

    pub fn decode(&mut self, input: &str) -> Result<Value> {
        let start = Instant::now();

        if input.trim().is_empty() {
            return Ok(Value::Null);
        }

        let mut lexer = Lexer::new(input);
        let tokens = lexer.tokenize();

        let mut parser = Parser::new(self.options.clone());
        let result = parser.parse(tokens)?;

        let elapsed = start.elapsed();
        let output_str = serde_json::to_string(&result)?;

        self.stats = NeonStats {
            input_size: input.len(),
            output_size: output_str.len(),
            decode_time_ms: elapsed.as_secs_f64() * 1000.0,
            ..Default::default()
        };

        Ok(result)
    }

    pub fn get_stats(&self) -> &NeonStats {
        &self.stats
    }
}

/// Decode a NEON string to JSON Value
pub fn decode(input: &str, options: Option<NeonDecodeOptions>) -> Result<Value> {
    let mut decoder = NeonDecoder::new(options);
    decoder.decode(input)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_decode_null() {
        let result = decode("N", None).unwrap();
        assert_eq!(result, Value::Null);
    }

    #[test]
    fn test_decode_boolean() {
        assert_eq!(decode("T", None).unwrap(), Value::Bool(true));
        assert_eq!(decode("F", None).unwrap(), Value::Bool(false));
    }

    #[test]
    fn test_decode_number() {
        assert_eq!(decode("42", None).unwrap(), Value::Number(42.into()));
        assert_eq!(decode("95K", None).unwrap(), Value::Number(95000.into()));
    }

    #[test]
    fn test_decode_string() {
        assert_eq!(
            decode("hello", None).unwrap(),
            Value::String("hello".to_string())
        );
        assert_eq!(
            decode("hello_world", None).unwrap(),
            Value::String("hello world".to_string())
        );
    }

    #[test]
    fn test_decode_array() {
        let result = decode("#3 1 2 3", None).unwrap();
        assert!(result.is_array());
        assert_eq!(result.as_array().unwrap().len(), 3);
    }
}
